File: ulohy/30-robozaba/reseni/main.py
from typing import List

def number_of_ways(target: int, current_position: int, available_jumps: List[int], memo: List[int]) -> int:
    result = 0

    for available_jump in available_jumps:
        next_position = current_position + available_jump

        if next_position > target:
            continue

        result += memo[next_position]

    return result


def read_int() -> int:
    return int(input())


if __name__ == "__main__":
    l = read_int()
    input()
    j = read_int()
    input()

    available_jumps = []
    for _ in range(j):
        available_jumps.append(read_int())

    memo = []
    for _ in range(l + 1):
        memo.append(0)
    memo[l] = 1

    for i in range(l - 1, -1, -1):
        memo[i] = number_of_ways(l, i, available_jumps, memo)

    print(memo[0])

File: ulohy/30-robozaba/reseni/main.js
const readline = require("readline");
const rl = readline.createInterface({
    input: process.stdin,
    output: process.stdout,
    terminal: false,
});

let linesRead = 0;
const lines = [];

async function readLine() {
    if (lines.length < 1) {
        for await (const line of rl) {
            lines.push(line);
        }
    }

    linesRead++;
    return lines[linesRead - 1];
}

main().then();

async function main() {
    const l = await readInt();
    await readLine();
    const j = await readInt();
    await readLine();

    const availableJumps = [];
    for (let i = 0; i < j; i++) {
        availableJumps.push(await readInt());
    }

    const memo = [];
    for (let i = 0; i < l + 1; i++) {
        memo.push(0);
    }

    memo[l] = 1;

    for (let i = memo.length - 2; i >= 0; i--) {
        memo[i] = numberOfWays(l, i, availableJumps, memo);
    }

    console.log(memo[0]);
}

function numberOfWays(target, currentPosition, availableJumps, memo) {
    let sum = 0;

    for (const availableJump of availableJumps) {
        const nextPosition = currentPosition + availableJump;

        if (nextPosition > target) {
            continue;
        }

        sum += memo[nextPosition];
    }

    return sum;
}

async function readInt() {
    return parseInt(await readLine());
}
